packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java

```diff
@@ -0,0 +1,142 @@
+   */
+  @Override
+  public Long setSurfaceProvider(@NonNull Long identifier) {
+    Preview preview = (Preview) instanceManager.getInstance(identifier);
+    flutterSurfaceTexture = textureRegistry.createSurfaceTexture();
+    SurfaceTexture surfaceTexture = flutterSurfaceTexture.surfaceTexture();
+    Preview.SurfaceProvider surfaceProvider =
+        new Preview.SurfaceProvider() {
+          @Override
+          public void onSurfaceRequested(SurfaceRequest request) {
+            surfaceTexture.setDefaultBufferSize(
+                request.getResolution().getWidth(), request.getResolution().getHeight());
+            Surface flutterSurface = cameraXProxy.createSurface(surfaceTexture);
+            request.provideSurface(
+                flutterSurface,
+                Executors.newSingleThreadExecutor(),
+                (result) -> {
+                  SystemServicesFlutterApiImpl systemServicesFlutterApi =
+                      cameraXProxy.createSystemServicesFlutterApi(binaryMessenger);
+                  int resultCode = result.getResultCode();
+                  switch (resultCode) {
+                    case SurfaceRequest.Result.RESULT_SURFACE_USED_SUCCESSFULLY:
+                      flutterSurface.release();
+                      break;
+                    case SurfaceRequest.Result.RESULT_REQUEST_CANCELLED:
+                    case SurfaceRequest.Result.RESULT_INVALID_SURFACE:
+                    case SurfaceRequest.Result.RESULT_WILL_NOT_PROVIDE_SURFACE:
+                      flutterSurface.release();
+                    case SurfaceRequest.Result.RESULT_SURFACE_ALREADY_PROVIDED:
+                    default:
+                      systemServicesFlutterApi.sendCameraError(
+                          getProvideSurfaceErrorDescription(resultCode), reply -> {});
+                      break;
+                  }
+                });
+          };
+        };
+    preview.setSurfaceProvider(surfaceProvider);
+    return flutterSurfaceTexture.id();
+  }
+
+  /**
+   * Returns an error description for each {@link SurfaceRequest.Result} that represents an error
+   * with providing a surface.
+   */
+  private String getProvideSurfaceErrorDescription(int resultCode) {
+    switch (resultCode) {
+      case SurfaceRequest.Result.RESULT_REQUEST_CANCELLED:
+        return "Provided surface was never attached to the camera becausethe SurfaceRequest was cancelled by the camera.";
+      case SurfaceRequest.Result.RESULT_INVALID_SURFACE:
+        return "Provided surface could not be used by the camera.";
+      case SurfaceRequest.Result.RESULT_SURFACE_ALREADY_PROVIDED:
+        return "Provided surface was never attached to the camera because the SurfaceRequest was cancelled by the camera.";
+      case SurfaceRequest.Result.RESULT_WILL_NOT_PROVIDE_SURFACE:
+        return "Surface was not attached to the camera because the SurfaceRequest was marked as 'will not provide surface'.";
+      default:
```

Comment:
Please include the passed in result code in this text. It helps a bunch with debugging.

Reply:
I still don't see the int in this string.

Reply:
The int is the `resultCode`. Open to suggestions on the format.

Reply:
I think I was looking at a non updated file. This looks good.